R/dataframe.R

```diff
@@ -157,6 +160,9 @@ as_data_frame.data.frame <- function(x, ...) {
 #' @export
 #' @rdname as_data_frame
 as_data_frame.list <- function(x, validate = TRUE, ...) {
+
+  x <- compact(x)
```

Comment:
I think it would be useful to explicitly state the invariant that this preserves, which (I think) is that all these operations should give the same result:

```R
data_frame(a = x, b = y)

as_data_frame(list(a = x, b = y))

df <- data_frame(a = x)
df$b <- y
```

Regardless of what the values of `x` and `y` are.